### PR TITLE
Optimize Semaphore::try_wake_one

### PR DESCRIPTION
**Pull request by matklad**
### What does this PR do?

Avoid double lookup in a hash-map by utilizing the [entry API](https://doc.rust-lang.org/stable/std/collections/struct.HashMap.html#method.entry).

### Motivation

Duplicate hashmap lookups are just a smell

### Related issues

Builds on #86 

### Checklist

[] I have added unit tests to the code I am submitting
[] My unit tests cover both failure and success scenarios
[] If applicable, I have discussed my architecture
[x] The new code I am adding is formatted using `rustfmt`

### CODE REVIEW

**Comment by glommer** [comment]
The funny thing about this is that I was just telling @HippoBaro the other day that it was unfortunate that Rust didn't support an API that would avoid the double lookup like C++ had.

Turns out I just wasn't aware of it (Maybe it is just me, but `entry` isn't the most obvious name for it...)

Good change then, because that is something that I was complaining about anyway.

**Comment by glommer** [comment]
There was a conflict with your previous PR to unify usage of `Future`.

It seemed trivial so I resolved it and as soon as CI comes back I'll merge this.

**Review by glommer — APPROVED** [review]

**Comment by matklad** [comment]
>It seemed trivial so I resolved it and as soon as CI comes back I'll merge this.

Btw, can I advertise using https://bors.tech/ which implements [not rocket science](https://graydon2.dreamwidth.org/1597.html) rule? It helps in particular in such situations, where you can just r+ and let the robot wait for CI. 

>Turns out I just wasn't aware of it (Maybe it is just me, but entry isn't the most obvious name for it...)

Yeah, this is one of the things you have to just know. There's a [crazier version](https://docs.rs/hashbrown/0.9.1/hashbrown/struct.HashMap.html#method.raw_entry_mut) in third party library.

**Pull request merged**
